packages/components/src/tabs/tablist.tsx

```diff
@@ -17,19 +23,115 @@ import type { TabListProps } from './types';
 import { useTabsContext } from './context';
 import { TabListWrapper } from './styles';
 import type { WordPressComponentProps } from '../context';
+import type { CSSProperties } from 'react';
+import classNames from 'classnames';
+
+function useTrackElementOffset(
+	targetElement?: HTMLElement | null,
+	onUpdate?: () => void
+) {
+	const [ indicatorPosition, setIndicatorPosition ] = useState( {
+		left: 0,
+		top: 0,
+		width: 0,
+		height: 0,
+	} );
+
+	// TODO: replace with useEventCallback or similar when officially available.
+	const updateCallbackRef = useRef( onUpdate );
+	useLayoutEffect( () => {
+		updateCallbackRef.current = onUpdate;
+	} );
+
+	const observedElementRef = useRef< HTMLElement >();
+	const resizeObserverRef = useRef< ResizeObserver >();
+	useEffect( () => {
+		if ( targetElement === observedElementRef.current ) return;
+		observedElementRef.current = targetElement ?? undefined;
+
+		function updateIndicator( element: HTMLElement ) {
+			setIndicatorPosition( {
+				left: element.offsetLeft,
+				top: element.offsetTop,
+				width: element.offsetWidth,
+				height: element.offsetHeight,
+			} );
+			updateCallbackRef.current?.();
```

Comment:
What difference does it make if I just call `onUpdate?.()` here (and add it as a dependency of course)? It's not immediately clear why we need the `updateCallbackRef` in the first place.

Reply:
Similar logic to https://github.com/WordPress/gutenberg/pull/60560#discussion_r1610112412, though in this case due to how the effect is implemented (idempotently) it wouldn't really break anything, but it would cause wasteful computation since there's an unnecessary dependency for the effect.